pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/RelianceOnDefaultCharsetRule.java

```diff
@@ -0,0 +1,109 @@
+        METHOD_TO_MIN_VERSION.put("java.io.PrintWriter#new(java.io.File)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.io.PrintWriter#new(java.io.OutputStream)", "10");
+        METHOD_TO_MIN_VERSION.put("java.io.PrintWriter#new(java.io.OutputStream,boolean)", "10");
+
+        METHOD_TO_MIN_VERSION.put("java.util.Scanner#new(java.io.InputStream)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Scanner#new(java.io.File)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Scanner#new(java.nio.file.Path)", "1.7");
+        METHOD_TO_MIN_VERSION.put("java.util.Scanner#new(java.nio.channels.ReadableByteChannel)", "1.10");
+
+        METHOD_TO_MIN_VERSION.put("java.util.Formatter#new()", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Formatter#new(java.io.OutputStream)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Formatter#new(java.io.PrintStream)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Formatter#new(java.lang.String)", "1.5");
+        METHOD_TO_MIN_VERSION.put("java.util.Formatter#new(java.io.File)", "1.5");
+        
+        METHOD_TO_MIN_VERSION.put("java.lang.String#new(byte[])", "1.6");
+        METHOD_TO_MIN_VERSION.put("java.lang.String#new(byte[],int,int)", "1.6");
+        METHOD_TO_MIN_VERSION.put("java.lang.String#getBytes()", "1.6");
+
+        METHOD_TO_MIN_VERSION.put("java.io.ByteArrayOutputStream#toString()", "10");
+        
+        METHOD_TO_MIN_VERSION.put("java.io.FileReader#new(java.lang.String)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileReader#new(java.io.File)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileReader#new(java.io.FileDescriptor)", "11");
+
+        METHOD_TO_MIN_VERSION.put("java.io.FileWriter#new(java.lang.String)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileWriter#new(java.lang.String,boolean)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileWriter#new(java.io.File)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileWriter#new(java.io.File,boolean)", "11");
+        METHOD_TO_MIN_VERSION.put("java.io.FileWriter#new(java.io.FileDescriptor)", "11");
+    }
+
+    public RelianceOnDefaultCharsetRule() {
+        super(ASTConstructorCall.class, ASTMethodCall.class);
+    }
+
+    @Override
+    public Object visit(ASTConstructorCall node, Object data) {
+        checkInvocation(node, data);
+        return data;
+    }
+
+    @Override
+    public Object visit(ASTMethodCall node, Object data) {
+        checkInvocation(node, data);
+        return data;
+    }
+
+    private void checkInvocation(net.sourceforge.pmd.lang.java.ast.JavaNode node, Object data) {
+        for (Map.Entry<String, String> entry : METHOD_TO_MIN_VERSION.entrySet()) {
+            String methodSignature = entry.getKey();
+            String minVersion = entry.getValue();
+            
+            InvocationMatcher matcher = InvocationMatcher.parse(methodSignature);
```

Comment:
this should really be avoided… I'd consider changing the map type to directly store `InvocationMatcher` in the keys (`InvocationMatcher` would require to implement `hashCode` / `equals`, or, since the map is always fully iterated, just have a list of a container tuple object…